lib/stats/stats-cluster.c

```diff
@@ -261,22 +289,29 @@ stats_counter_group_init_equals(const StatsCounterGroupInit *self, const StatsCo
 gboolean
 stats_cluster_key_equal(const StatsClusterKey *key1, const StatsClusterKey *key2)
 {
-  return key1->legacy.component == key2->legacy.component
-         && strcmp(key1->id, key2->id) == 0
-         && strcmp(key1->legacy.instance, key2->legacy.instance) == 0
-         && stats_counter_group_init_equals(&key1->counter_group_init, &key2->counter_group_init)
-         && stats_cluster_key_labels_equal(key1->labels, key1->labels_len, key2->labels, key2->labels_len);
+  if (stats_cluster_key_is_legacy(key1) != stats_cluster_key_is_legacy(key2))
+    return FALSE;
+
+  if (stats_cluster_key_is_legacy(key1) && stats_cluster_key_is_legacy(key2))
+    {
+      return key1->legacy.component == key2->legacy.component
+             && strcmp(key1->legacy.id, key2->legacy.id) == 0
+             && strcmp(key1->legacy.instance, key2->legacy.instance) == 0
+             && stats_counter_group_init_equals(&key1->counter_group_init, &key2->counter_group_init);
+    }
+
+  return strcmp(key1->name, key2->name) == 0
+         && stats_cluster_key_labels_equal(key1->labels, key1->labels_len, key2->labels, key2->labels_len)
+         && stats_counter_group_init_equals(&key1->counter_group_init, &key2->counter_group_init);
```

Comment:
This function raised some questions for me.

It seems like the implementation only accounts for cleanly new-style and cleanly legacy-style keys. By this I mean if both of them are legacy, only the legacy parts are compared, if they are both new-style, only the new-style values are checked.

If I understand correctly there is a third case, which is new-style with legacy alias, where all of the fields should be checked.

I wonder what should happen, if two keys have the same new-style values, but different legacy aliases or vice versa. Should we even allow that? If not, how should we implement that, so it does not get really slow?

If I understand correctly, currently if there are two keys with different new-style values and the same alias, they get the same cluster with the new-style values of the first one to register. I don't think this is okay.

Reply:
I think it will be easier to remove the legacy code this way. We should not use different new-style keys with the same legacy alias.

Reply:
Agreed.